### PR TITLE
Build against Spring Security

### PR DESCRIPTION
**Pull request by jglick**
Following up #48 but now trying to build against https://github.com/jenkinsci/jenkins/pull/4848 and use only Spring Security.

Managed to make everything compile & pass tests. @rwinch I will likely need your help here. There seem to have been quite substantial changes between Acegi Security and Spring Security in the LDAP support, and I have not managed to find a migration guide. Most (though not all) of the problems center around some change from using `LdapUserDetails`, which used to have a `getAttributes` method, to `DirContextOperations` (which has this method) plus a `String username` but which is not an `Authentication` or `UserDetails`. I am feeling my way around in the dark, knowing next to nothing about LDAP, JNDI, etc.

### CODE REVIEW

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:907]: Not really sure what the Spring Security implementation is going to throw instead of this, so just guessing.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:901]: The most appropriate-sounding type I could find in general.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:1148]: Not sure what this was, but anyway it is gone now.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:1310]: Here is one of several places where I could not figure out how to get things to even compile. If I change the type of the variable to `DirContextOperations` then some errors go away (calls to `getGrantedAuthorities` and `getAttributes`) but many others appear.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:1498]: Awkwardly, the supertype has

```java
	/**
	 * Returns the default role Method available so that classes extending this can
	 * override
	 * @return the default role used
	 * @see #setDefaultRole(String)
	 */
	private GrantedAuthority getDefaultRole() {
		return this.defaultRole;
	}
```

which does not make sense—it is _not_ accessible.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [inline src/main/java/hudson/security/LDAPSecurityRealm.java]: `protected final` in supertype.

File: src/main/java/jenkins/security/plugins/ldap/BindAuthenticator2.java
Comment (jglick) [inline src/main/java/jenkins/security/plugins/ldap/BindAuthenticator2.java]: I guess this is the replacement??

File: src/main/java/jenkins/security/plugins/ldap/BindAuthenticator2.java
Comment (jglick) [inline src/main/java/jenkins/security/plugins/ldap/BindAuthenticator2.java]: `handleBindException` is now `protected` so no more need to align packages.

File: src/main/java/jenkins/security/plugins/ldap/FromGroupSearchLDAPGroupMembershipStrategy.java
Comment (jglick) [inline src/main/java/jenkins/security/plugins/ldap/FromGroupSearchLDAPGroupMembershipStrategy.java:81]: Just trying to make something work.

File: src/main/java/jenkins/security/plugins/ldap/FromGroupSearchLDAPGroupMembershipStrategy.java
Comment (jglick) [inline src/main/java/jenkins/security/plugins/ldap/FromGroupSearchLDAPGroupMembershipStrategy.java:88]: Just guessing here. Read only? Read write? `retrieveEntry` is nowhere to be found in Spring Security so I tried to inline the Acegi Security behavior to the extent I understood it. Note that this means `GroupMembersMapper.mapAttributes` could just be inlined here too, but I am trying to keep the diff small where possible.

File: src/main/java/jenkins/security/plugins/ldap/FromUserRecordLDAPGroupMembershipStrategy.java
Comment (jglick) [inline src/main/java/jenkins/security/plugins/ldap/FromUserRecordLDAPGroupMembershipStrategy.java:112]: Maybe needs to be passed in as an additional argument or something?

File: src/main/java/jenkins/security/plugins/ldap/LDAPConfiguration.java
Comment (jglick) [inline src/main/java/jenkins/security/plugins/ldap/LDAPConfiguration.java]: Just guessing.

File: src/main/java/jenkins/security/plugins/ldap/LDAPConfiguration.java
Comment (jglick) [inline src/main/java/jenkins/security/plugins/ldap/LDAPConfiguration.java]: Inferred from `LdapAuthenticationProviderConfigurer` that this is the equivalent.

File: src/main/java/jenkins/security/plugins/ldap/LDAPConfiguration.java
Comment (jglick) [inline src/main/java/jenkins/security/plugins/ldap/LDAPConfiguration.java]: Could not find any equivalent to this method in `DefaultInitialDirContextFactory`.

Anyway I think `extraEnvVars` is deprecated in the Jenkins plugin. Not sure about the JNDI properties.

File: src/main/java/jenkins/security/plugins/ldap/LDAPExtendedTemplate.java
Comment (jglick) [inline src/main/java/jenkins/security/plugins/ldap/LDAPExtendedTemplate.java:103]: Again, read only? Read write?

File: src/main/java/jenkins/security/plugins/ldap/LdapEntryMapper.java
Comment (jglick) [inline src/main/java/jenkins/security/plugins/ldap/LdapEntryMapper.java:33]: Could find no exact equivalent of this Acegi Security type, but anyway it is only being used internally here.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (Vlatombe) [reply src/main/java/hudson/security/LDAPSecurityRealm.java:1310]: I've been looking into this. `LdapUserDetailsManager` seems to be a replacement for the whole method.

However, we would need a custom `UserDetailsContextMapper` in order to map `LDAPConfiguration#displayNameAttributeName` and `LDAPConfiguration#mailAddressAttributeName`

File: src/main/java/jenkins/security/plugins/ldap/FromGroupSearchLDAPGroupMembershipStrategy.java
Comment (Vlatombe) [reply src/main/java/jenkins/security/plugins/ldap/FromGroupSearchLDAPGroupMembershipStrategy.java:88]: All the calls should be read-only. The ldap plugin never writes in the remote ldap afair.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [reply src/main/java/hudson/security/LDAPSecurityRealm.java:1310]: Great, I will look into that.

Comment (jglick) [reply src/main/java/hudson/security/LDAPSecurityRealm.java]: Looked at it, but that seemed to be a way bigger refactoring, and I did not see how to fit it into the plugin’s current logic. For now, went with a simpler patch that at least fixes the types here.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:1339]: Is this right?

**Comment by rsandell** [comment]
Since this will be a major version change with long sought after upgrades, I think removing some of the small obscure features is OK.

**Comment by jglick** [comment]
> I think removing some of the small obscure features is OK.

Probably so, but I do not know enough about this plugin’s usage to even know which features are “small” or “obscure”.

Currently there are still quite a few test failures that are beyond the limits of my knowledge. For example: `ou=umich,ou.edu` is apparently an “invalid name” according to JNDI. This is just the search string (or “URL” or something?) configured in the test, but apparently it is supposed to be processed somehow?

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:1298]: Could probably clean some of this stuff up in `master` at this point, in a separate PR.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:1329]: If this cache thing is significant, need to revisit.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:1412]: Not covered by tests I guess?

File: src/test/java/hudson/security/LDAPEmbeddedTest.java
Comment (jglick) [inline src/test/java/hudson/security/LDAPEmbeddedTest.java:551]: Spent most of the day trying to figure this out, without success. Not clear whether there is actually something wrong in the plugin, or just the test making unwarranted assumptions.

File: src/test/java/hudson/security/LDAPSecurityRealmTest.java
Comment (jglick) [inline src/test/java/hudson/security/LDAPSecurityRealmTest.java:57]: Does not even compile IIRC, and did not seem critical.

**Comment by jglick** [comment]
Hmm, while functional tests pass, most ATH tests are failing with

```
java.lang.ClassCastException: com.sun.jndi.ldap.LdapCtx cannot be cast to org.springframework.ldap.core.DirContextAdapter
	at org.springframework.security.ldap.SpringSecurityLdapTemplate.searchForSingleEntryInternal(SpringSecurityLdapTemplate.java:337)
	at org.springframework.security.ldap.SpringSecurityLdapTemplate.lambda$searchForSingleEntry$2(SpringSecurityLdapTemplate.java:312)
	at org.springframework.ldap.core.LdapTemplate.executeWithContext(LdapTemplate.java:817)
	at org.springframework.ldap.core.LdapTemplate.executeReadOnly(LdapTemplate.java:803)
	at org.springframework.security.ldap.SpringSecurityLdapTemplate.searchForSingleEntry(SpringSecurityLdapTemplate.java:312)
	at org.springframework.security.ldap.search.FilterBasedLdapUserSearch.searchForUser(FilterBasedLdapUserSearch.java:128)
	at org.springframework.security.ldap.authentication.BindAuthenticator.authenticate(BindAuthenticator.java:95)
	at jenkins.security.plugins.ldap.BindAuthenticator2.authenticate(BindAuthenticator2.java:50)
	at org.springframework.security.ldap.authentication.LdapAuthenticationProvider.doAuthentication(LdapAuthenticationProvider.java:187)
	at org.springframework.security.ldap.authentication.AbstractLdapAuthenticationProvider.authenticate(AbstractLdapAuthenticationProvider.java:85)
	at org.springframework.security.authentication.ProviderManager.authenticate(ProviderManager.java:199)
	at hudson.security.LDAPSecurityRealm$LDAPAuthenticationManager.authenticate(LDAPSecurityRealm.java:1008)
	at org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter.attemptAuthentication(UsernamePasswordAuthenticationFilter.java:95)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:212)
	at …
```

The failing cast (from Object) is uncommented:

```java
DirContextAdapter dca = (DirContextAdapter) searchResult.getObject();
```

Seems there are two only partially related inheritance hierarchies under `DirContext`. All cryptic to me.

**Comment by jglick** [comment]
Reproducible also in `hpi:run`. In `DirectoryManager`,

```java
        FactoryEnumeration factories = ResourceManager.getFactories(
            Context.OBJECT_FACTORIES, environment, nameCtx);
```

produces `null` despite the `environment` specifying `java.naming.factory.object` as `org.springframework.ldap.core.support.DefaultDirObjectFactory`. Smells like a class loading issue.

**Comment by jglick** [comment]
I think I know what to fix here; just having trouble running ATH locally to confirm it.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (rsandell) [reply src/main/java/hudson/security/LDAPSecurityRealm.java:1330]: No clue, but if it works it works right? :/

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (rsandell) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:1524]: Why?

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (rsandell) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:1773]: The old `LdapDataAccessException` meant that there was problems communicating with the ldap server, or your query was wrong or similar. Is the new `AuthenticationException` equivalent? By the name it sounds to me like it's a different exception.

File: src/main/java/jenkins/security/plugins/ldap/BindAuthenticator2.java
Comment (rsandell) [reply src/main/java/jenkins/security/plugins/ldap/BindAuthenticator2.java:45]: seems like it...

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [reply src/main/java/hudson/security/LDAPSecurityRealm.java:1524]: Oh this was a bit of a mess indeed. See https://github.com/jenkinsci/ldap-plugin/pull/49#discussion_r463740875 amended by https://github.com/jenkinsci/ldap-plugin/pull/49/commits/a9dc4e48687f07fdf4ca374d55cb4ac662f6d425 (which also reduces diff size). Basically the plugin never used the `super` implementation of `convertToUpperCase` and `rolePrefix`; instead https://github.com/jenkinsci/ldap-plugin/blob/9577fe6b8ea685d04fcfe243e750eeea059a88a2/src/main/java/hudson/security/LDAPSecurityRealm.java#L1506-L1515 reimplements something like it, for reasons I do not entirely follow. So to disable the `super` implementations, but still capture values passed to the setters, while accommodating a supertype which newly made the getters (but not setters) `final`, we have these tricks.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [reply src/main/java/hudson/security/LDAPSecurityRealm.java:1773]: There is no more `org.springframework.dao.DataAccessException` or its subtypes, so in the core PR as well as here I have converted these to the more generic `AuthenticationException`. Whether this is the most appropriate type or not, I do not know, and I am not sure we really care.

File: src/main/java/jenkins/security/plugins/ldap/FromUserRecordLDAPGroupMembershipStrategy.java
Comment (rsandell) [reply src/main/java/jenkins/security/plugins/ldap/FromUserRecordLDAPGroupMembershipStrategy.java:112]: isn't it the second argument to the method you are in?

File: src/test/java/hudson/security/LDAPEmbeddedTest.java
Comment (jglick) [reply src/test/java/hudson/security/LDAPEmbeddedTest.java:551]: also https://github.com/jenkinsci/acceptance-test-harness/pull/607

Comment (rsandell) [reply src/test/java/hudson/security/LDAPEmbeddedTest.java:551]: I'm actually not sure what this test is supposed to test, except judging by the name of the test it checks that environment properties are set correctly. So I don't think it specifically needs to check that you can't login after setting `java.naming.ldap.typesOnly` :) It could test some other value instead.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (rsandell) [inline src/main/java/hudson/security/LDAPSecurityRealm.java:745]: History lesson for context;
When we implemented the multiple ldap server support we were a bit paranoid of changing the behavior in any way for configurations with only one server configured. That's why you see all of these `if (configurations.size() > 1)` all over the place. It was also the reason for the `DelegatedLdapUserDetails`  class that I see you have changed to always be used instead of before; only when multiple configurations are in effect.

Since the behavior _(both user wise but also in binary compatibility)_ is changing now and `DelegatedLdapUserDetails` is used all the time, I believe it's now possible to remove all this semi redundant code that has separate treatments for the single server configuration scenario.

But I won't block this PR if you don't want to include that cleanup work here ;).

File: src/test/java/hudson/security/LdapMultiEmbeddedTest.java
Comment (rsandell) [inline src/test/java/hudson/security/LdapMultiEmbeddedTest.java:406]: I think the original intent here was to check that there actually was two records in the log.

Comment (jglick) [reply src/test/java/hudson/security/LdapMultiEmbeddedTest.java:406]: Perhaps that was the intent, but in fact it was just repeating an identical assertion.

Comment (jglick) [reply src/test/java/hudson/security/LdapMultiEmbeddedTest.java:406]: See original https://github.com/jenkinsci/ldap-plugin/blob/ff2c95a970df0c0446fe31d3b218c8bded827c2e/src/test/java/hudson/security/LdapMultiEmbeddedTest.java#L401-L410 which AFAICT was just a mistake.

Comment (jglick) [reply src/test/java/hudson/security/LdapMultiEmbeddedTest.java:406]: See #31 by @dwnusbaum. `when_first_and_second_are_unavailable_and_ignorable_then_no_login` sensibly verifies that we have messages from both `planetExpress` and `sevenSeas`. Yet `when_first_and_second_are_unavailable_and_ignorable_then_lookup_fails`, perhaps a modified copy of that test, redundantly asserts that we have a message from `planetExpress` only.

Comment (dwnusbaum) [reply src/test/java/hudson/security/LdapMultiEmbeddedTest.java:406]: I think it was a typo, the second assertion should have used `sevenSeas.getPort` to make sure that the plugin tried to check both servers.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [reply src/main/java/hudson/security/LDAPSecurityRealm.java:744]: Trying this in #52.

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [reply src/main/java/hudson/security/LDAPSecurityRealm.java:1300]: ⇒ #52

File: src/main/java/hudson/security/LDAPSecurityRealm.java
Comment (jglick) [reply src/main/java/hudson/security/LDAPSecurityRealm.java]: see also https://github.com/jenkinsci/ldap-plugin/pull/49#discussion_r495062096

File: src/main/java/jenkins/security/plugins/ldap/FromGroupSearchLDAPGroupMembershipStrategy.java
Comment (jglick) [reply src/main/java/jenkins/security/plugins/ldap/FromGroupSearchLDAPGroupMembershipStrategy.java:87]: > could just be inlined here too

⇒ #52

**Review by rsandell — APPROVED** [review]

**Review by rsandell — APPROVED** [review]

File: pom.xml
Comment (jvz) [inline pom.xml:31]: To be released in the next weekly. Will this be ready to review and merge afterward?

Comment (jglick) [reply pom.xml:31]: It has already been reviewed (more reviews welcome!) and yes it would be ready to merge when the dep is on 2.266.

**Pull request merged**
